scripts/build_desktop_app_with_firebase.py

```diff
@@ -0,0 +1,281 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Standalone self-sufficient (no external dependencies) python script that can ease
+building desktop apps depending on firebase, by either using the firebase cpp
+source (firebase-cpp-sdk github repo) or the prebuilt release firebase libraries.
+
+Note that this script works with only Python3 (3.6+).
+Known side effects:
+If building against firebase cpp source, this script might checkout a specific
+branch on github containing vcpkg. This will not be required once vcpkg is in the
+main branch.
+
+Example usage:
+Let's say we want to build the quickstart cpp example for firebase database.
+As specified above, there are 2 options - build against the firebase source or
+prebuilt firebase libraries.
+
+# Build against the firebase cpp sdk source
+python3 scripts/build_desktop_app_with_firebase.py --app_dir ~/quickstart-cpp/database/testapp
+                                                   --sdk_dir . --build_dir build_source
+
+(or)
+
+# Build against the prebuilt released firebase libraries
+python3 scripts/build_desktop_app_with_firebase.py --app_dir ~/quickstart-cpp/database/testapp
+                                                   --sdk_dir ~/prebuilt/firebase_cpp_sdk_6.15.1/
+                                                   --build_dir build_prebuilt
+
+# If the script ran successfully, it will print the path to the build directory.
+Build successful!
+Please find your executables in build directory: /Users/<user>/quickstart-cpp/database/testapp/build_source
+
+# Running the built example
+$ ./Users/<user>/quickstart-cpp/database/testapp/build_source/desktop_testapp
+"""
+import argparse
+import distutils.spawn
+import os
+import platform
+import subprocess
+import sys
+
+def is_path_valid_for_cmake(path):
+  """Check if specified path is setup for cmake"""
+  return os.path.exists(os.path.join(path, 'CMakeLists.txt'))
+
+def is_sdk_path_source(sdk_dir):
+  """Validate if firebase sdk dir is firebase cpp source dir"""
+  # Not the most reliable way to detect if the sdk path is source or prebuilt but
+  # should work for our purpose.
+  return os.path.exists(os.path.join(sdk_dir, 'build_tools'))
+
+def validate_prebuilt_args(arch, config):
+  """Validate cmd line args for build with prebuilt libraries"""
+  # Some options are not available when using prebuilt libraries"""
+  if platform.system() == 'Darwin':
+    if arch == 'x86' or config == 'Debug':
+      raise ValueError("Prebuilt mac firebase libraries are built for x64 and Release mode only. "
+                       "Please fix the command line arguments and try again")
+
+  if platform.system() == 'Linux':
+    if config == 'Debug':
+      raise ValueError("Prebuilt linux firebase libraries are built with Release mode only. "
+                       "Please fix the --config command line argument and try again.")
+
+def get_vcpkg_triplet(arch, msvc_runtime_library='static'):
```

Comment:
This logic (and some others) already exist in the SDK repo, don't they? Should we find a way to avoid duplication?

Reply:
Good point. We should at some point consolidate these scripts but I feel the main goal here was to provide a super simple self sufficient single script for our users to understand how to build things with vcpkg. If the script does too many other things, the logic might not be that straightforward to follow. 
I wouldn't want to add integration test specific command line arguments (eg: api) to this script (just to keep this one generic that users can place anywhere) but we could call this script from integration scripts as that seems to be a more special case.

Reply:
Marking this as resolved for the time being but we should definitely make a note of this.